File: src/residue_native.cc
#include <nan.h>
#include <residue/residue.h>

using v8::FunctionTemplate;
using v8::Handle;
using v8::Object;
using v8::Isolate;
using v8::String;
using Nan::To;
using Nan::GetFunction;
using Nan::New;
using Nan::Set;

NAN_METHOD(Version)
{
    info.GetReturnValue().Set(String::NewFromUtf8(Isolate::GetCurrent(), Residue::version().c_str()));
}

NAN_METHOD(Configure)
{
    String::Utf8Value jsonParam(info[0]->ToString());
    std::string json(*jsonParam);
    Residue::loadConfigurationFromJson(json);
}

NAN_METHOD(LoadConnection)
{
    String::Utf8Value jsonParam(info[0]->ToString());
    std::string json(*jsonParam);
    Residue::loadConnectionFromJson(json);
}

NAN_METHOD(Connect)
{
    Residue::reconnect();
}

#define DEFINE_LOG_FN(name) NAN_METHOD(name) {       \
String::Utf8Value paramLoggerId(info[0]->ToString());    \
String::Utf8Value paramFile(info[1]->ToString());        \
String::Utf8Value paramLine(info[2]->ToString());        \
String::Utf8Value paramFunc(info[3]->ToString());        \
String::Utf8Value paramMsg(info[4]->ToString());         \
String::Utf8Value paramVLevel(info[5]->ToString());      \
std::string loggerId(*paramLoggerId);                    \
std::string file(*paramFile);                            \
std::string line(*paramLine);                            \
std::string func(*paramFunc);                            \
std::string msg(*paramMsg);                              \
std::string vlevel(*paramVLevel);                        \
el::base::type::LineNumber lineNumb = 0;                 \
if (line != "undefined" && line != "null") {             \
    lineNumb = stoi(line);                               \
}                                                        \
el::base::type::VerboseLevel vl = 0;                     \
if (vlevel != "undefined" && vlevel != "null") {         \
    vl = stoi(vlevel);                                   \
}                                                        \
el::base::Writer(el::Level::name, file.c_str(), lineNumb, func.c_str(), el::base::DispatchAction::NormalLog, vl).\
    construct(1, loggerId.c_str()) << msg;\
}

DEFINE_LOG_FN(Info)
DEFINE_LOG_FN(Error)
DEFINE_LOG_FN(Trace)
DEFINE_LOG_FN(Debug)
DEFINE_LOG_FN(Warning)
DEFINE_LOG_FN(Verbose)

#undef DEFINE_LOG_FN

NAN_MODULE_INIT(InitAll)
{
    #define DEFINE_FN(name, callback) Set(target, New<String>(#name).ToLocalChecked(), \
        GetFunction(New<FunctionTemplate>(callback)).ToLocalChecked())

    DEFINE_FN(version, Version);
    DEFINE_FN(configure, Configure);
    DEFINE_FN(load_connection, LoadConnection);
    DEFINE_FN(connect, Connect);

    DEFINE_FN(info, Info);
    DEFINE_FN(error, Error);
    DEFINE_FN(trace, Trace);
    DEFINE_FN(debug, Debug);
    DEFINE_FN(warning, Warning);
    DEFINE_FN(verbose, Verbose);

    #undef DEFINE_FN
}


NODE_MODULE(residue_native, InitAll)

File: src/residue-native.js

const util = require('util');
const residue_native = require('./build/Release/residue-native-binding');

const getSourceLocation = (splitChar) => (new Error).stack.split('\n')[5].replace(' at ', '').trim().split(splitChar);

const getSourceFile = () => getSourceLocation(':')[0];

const getSourceLine = () => parseInt(getSourceLocation(':')[1]);

const getSourceFunc = () => {
    const parts = getSourceLocation(' ');
    if (parts.length <= 1) {
        return 'anonymous';
    }
    return parts[0];
}

const Residue = function() {
    this.version = () => {
        return residue_native.version();
    }

    this.configure = (json) => {
        residue_native.configure(json);
    }

    this.connect = () => {
        residue_native.connect();
    }
};

const Logger = function(id) {
    this.id = id;

    this.info = (fmt, ...args) => {
        this._send_log_msg(residue_native.info, undefined, fmt, ...args);
    }

    this.error = (fmt, ...args) => {
        this._send_log_msg(residue_native.error, undefined, fmt, ...args);
    }

    this.debug = (fmt, ...args) => {
        this._send_log_msg(residue_native.debug, undefined, fmt, ...args);
    }

    this.warning = (fmt, ...args) => {
        this._send_log_msg(residue_native.warning, undefined, fmt, ...args);
    }

    this.trace = (fmt, ...args) => {
        this._send_log_msg(residue_native.trace, undefined, fmt, ...args);
    }

    this.verbose = (vlevel, fmt, ...args) => {
        this._send_log_msg(residue_native.verbose, vlevel, fmt, ...args);
    }

    // private
    this._send_log_msg = (fn, vlevel, fmt, ...args) => {
        const cpy = args;
        for (var idx = 0; idx < cpy.length; ++idx) {
            if (typeof cpy[idx] === 'object') {
                cpy[idx] = JSON.stringify(cpy[idx]);
            }
        }

        fn(this.id, getSourceFile(), getSourceLine(), getSourceFunc(), util.format(fmt, ...cpy), vlevel);
    }
};

exports.Residue = Residue;
exports.Logger = Logger;
